qvisor/src/kvm_vcpu_aarch64.rs

```diff
@@ -168,42 +163,42 @@ impl KVMVcpu {
-                            true
-                        });
-
                         panic!("kvm virtual cpu[{}] run failed: Error {:?}", self.id, e)
                     }
                 }
             };
-            self.state
-            .store(KVMVcpuState::HOST as u64, Ordering::Release);
+            self.state.store(KVMVcpuState::HOST as u64, Ordering::Release);
             match kvmRet {
-                VcpuExit::MmioRead(addr, data) => {
+                VcpuExit::MmioRead(addr, _data) => {
+                    self.backtrace()?;
                     panic!(
                         "CPU[{}] Received an MMIO Read Request for the address {:#x}.",
                         self.id, addr,
                     );
                 }
-                VcpuExit::MmioWrite(addr, data) => {
+                VcpuExit::MmioWrite(addr, _data) => {
                     {
                         let mut interrupting = self.interrupting.lock();
                         interrupting.0 = false;
                         interrupting.1.clear();
                     }
-                    let para1 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R0).map_err(|e| Error::SysError(e.errno()))?;
-                    let para2 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R1).map_err(|e| Error::SysError(e.errno()))?;
-                    let para3 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R2).map_err(|e| Error::SysError(e.errno()))?;
-                    let para4 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R3).map_err(|e| Error::SysError(e.errno()))?;
-                    if addr > (u16::MAX as u64) {
-                        panic!("cpu[{}] Received hypercall id max than 255");
+                    // reading hypercall parameters from vcpu register file
+                    // x0 and x1 (w1) are used by the str instruction
+                    // the 64-bit parameters 1,2,3,4 are passed via
+                    // x2,x3,x4,x5
+                    let para1 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R2).map_err(|e| Error::SysError(e.errno()))?;
+                    let para2 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R3).map_err(|e| Error::SysError(e.errno()))?;
+                    let para3 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R4).map_err(|e| Error::SysError(e.errno()))?;
+                    let para4 = self.vcpu.get_one_reg(_KVM_ARM64_REGS_R5).map_err(|e| Error::SysError(e.errno()))?;
+                    let hypercall_id = (addr - MemoryDef::HYPERCALL_MMIO_BASE) as u16;
+                    if hypercall_id > u16::MAX {
+                        panic!("cpu[{}] Received hypercall id max than 255", self.id);
                     }
-                    self.handle_hypercall(addr as u16, data, para1, para2, para3, para4)?;
+                    // logging all hypercalls can be very noisy. Disabled by default.
+                    // debug!("hypecall: {}, paras: {:x} {:x} {:x} {:x}", hypercall_id, para1, para2, para3, para4);
```

Comment:
Maybe we can do this through the config and/or new debug macro for HCalls (in the future).

Reply:
Good point. Please take notes somewhere, I'm resolving this one for clarity in this PR.